services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisEnforcerExecutor.java

```diff
@@ -0,0 +1,516 @@
+            throw invalidRedirectUri(ERR_GENERAL);
+        }
+ 
+        validation.validate();
+    }
+
+    public static class UriValidation {
+        public final URI uri;
+        public final boolean isRedirectUriParameter;
+        public final Configuration config;
+
+        public UriValidation(String uriString, boolean isRedirectUriParameter, Configuration config) throws URISyntaxException {
+            this.uri = new URI(uriString);
+            this.isRedirectUriParameter = isRedirectUriParameter;
+            this.config = config;
+        }
+
+        public void validate() throws ClientPolicyException {
+            switch (identifyUriType()) {
+                case IPV4_LOOPBACK_ADDRESS:
+                    if(!config.isAllowIPv4LoopbackAddress() || !isValidIPv4LoopbackAddress()) {
+                        throw invalidRedirectUri(ERR_LOOPBACK);
+                    }
+                    break;
+                case IPV6_LOOPBACK_ADDRESS:
+                    if(!config.isAllowIPv6LoopbackAddress() || !isValidIPv6LoopbackAddress()) {
+                        throw invalidRedirectUri(ERR_LOOPBACK);
+                    }
+                    break;
+                case PRIVATE_USE_URI_SCHEME:
+                    if(!config.isAllowPrivateUseUriScheme() || !isValidPrivateUseUriScheme()) {
+                        throw invalidRedirectUri(ERR_PRIVATESCHEME);
+                    }
+                    break;
+                case NORMAL_URI:
+                    if(!isValidNormalUri()) {
+                        throw invalidRedirectUri(ERR_NORMALURI);
+                    }
+                    break;
+                default :
+                    logger.debugv("Invalid URI Type - input = {0}", uri.toString());
+                    throw invalidRedirectUri(ERR_GENERAL);
+            }
+        }
+
+        UriType identifyUriType() {
+            // NOTE: the order of evaluation methods is important.
+            if (isIPv4LoopbackAddress()) {
+                return UriType.IPV4_LOOPBACK_ADDRESS;
+            } else if (isIPv6LoopbackAddress()) {
+                return UriType.IPV6_LOOPBACK_ADDRESS;
+            } else if (isPrivateUseScheme()) {
+                return UriType.PRIVATE_USE_URI_SCHEME;
+            } else if (isNormalUri()) {
+                return UriType.NORMAL_URI;
+            } else {
+                return UriType.INVALID_URI;
+            }
+        }
+
+        boolean isIPv4LoopbackAddress() {
+            return isLoopbackAddressRedirectUri(i->i instanceof Inet4Address);
+        }
+
+        boolean isIPv6LoopbackAddress() {
+            return isLoopbackAddressRedirectUri(i->i instanceof Inet6Address);
+        }
+
+        boolean isLoopbackAddressRedirectUri(Predicate<InetAddress> p) {
+            if (uri.getHost() == null) {
+                return false;
+            }
+
+            InetAddress addr;
+            try {
+                addr = InetAddress.getByName(uri.getHost());
```

Comment:
+1

Reply:
I fixed the point as you suggested.

Reply:
I reverted the code because it causes an exception when the address of the URI is IPv6 address surrounded by square brackets which can be accepted by this executor.

Reply:
@tnorimat I think that with ipv6 the brackets are compulsory in URIs, but I can execute the test with this changes:

```diff
diff --git a/services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisEnforcerExecutor.java b/services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisEnforcerExecutor.java
index c6015299e9..c0c5a6d2e9 100644
--- a/services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisEnforcerExecutor.java
+++ b/services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisEnforcerExecutor.java
@@ -17,12 +17,12 @@
 
 package org.keycloak.services.clientpolicy.executor;
 
+import com.google.common.net.InetAddresses;
 import java.net.Inet4Address;
 import java.net.Inet6Address;
 import java.net.InetAddress;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.net.UnknownHostException;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
@@ -298,8 +298,8 @@ public class SecureRedirectUrisEnforcerExecutor implements ClientPolicyExecutorP
 
             InetAddress addr;
             try {
-                addr = InetAddress.getByName(uri.getHost());
-            } catch (UnknownHostException e) {
+                addr = InetAddresses.forUriString(uri.getHost());
+            } catch (IllegalArgumentException e) {
                 return false;
             }
```

Or is another test the one that fails?

Reply:
@rmartinc Yes. It worked well. Thank you very much!